BUG: only tell/seek np.fromfile when buffered

cjermain opened the pull request:
This fixes bug #6246 in Python 3, by using the `io` module to check if the file object is a child of the [`RawIOBase`](https://docs.python.org/3/library/io.html#io.RawIOBase), which are always unbuffered. Since it is not buffered, there is no reason to call tell or seek, which can fail on some legitimate file objects (e.g. wking/pycomedi#3).

cjermain committed:
BUG: only require tell/seek in np.fromfile when buffered

mhvk commented:
Wouldn't it make more sense to duck-type this, i.e., only check whether the file is unbuffered when `tell` has failed? That avoids needless work for the likely more common case where it is a regular file.

cjermain commented:
Yea, that makes more sense. Will fix...

pv commented:
tell() doesn't fail for buffered files, the Python-side buffering just
ends up in an inconsistent state, so I don't think the check can be avoided.

cjermain commented:
How about that?

cjermain commented:
Ok. How about that? I can squash the commits into 1 if needed.

mhvk commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -193,9 +193,26 @@ npy_PyFile_Dup2(PyObject *file, char *mode, npy_off_t *orig_pos)
+        /* File object instances of RawIOBase are unbuffered */
+        io_raw = PyObject_GetAttrString(io, "RawIOBase");
+        if (io_raw == NULL) {
+            return NULL;
```
This is not really my forte at all, but shouldn't `io` be decref'd here?

mhvk commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -241,9 +258,28 @@ npy_PyFile_DupClose2(PyObject *file, FILE* handle, npy_off_t orig_pos)
+        /* File object instances of RawIOBase are unbuffered */
+        io_raw = PyObject_GetAttrString(io, "RawIOBase");
+        if (io_raw == NULL) {
+            return -1;
```
Again, should one decref `io`?

cjermain commented:
If [`Py_DECREF` must not be called on `NULL`](https://docs.python.org/3.5/c-api/refcounting.html#c.Py_DECREF), so I don't think they apply.  Also see the `os` import in the previous part of that file which does not. I did make a minor fix on the positioning of the `Py_DECREF` statements for `io`.

mhvk commented:
@cjermain - The fix you did is what I had in mind; I was worried one might be left with a hanging reference to `io` itself.

cjermain commented:
I've rebased this as a single commit. Anything else needed?

charris commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -193,9 +193,26 @@ npy_PyFile_Dup2(PyObject *file, char *mode, npy_off_t *orig_pos)
+        }
+        unbuf = PyObject_IsInstance(file, io_raw);
+        Py_DECREF(io_raw);
+        if (unbuf == 1) { // Unbuffered
```
No C++ comments. Put on next line with `/*...*/`

charris commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -193,9 +193,26 @@ npy_PyFile_Dup2(PyObject *file, char *mode, npy_off_t *orig_pos)
+        Py_DECREF(io_raw);
+        if (unbuf == 1) { // Unbuffered
+            return handle;
+        } else {
```
This style:
```
}
else {
```
See `doc/C_STYLE_GUIDE.rst.txt`.

charris commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -241,9 +258,28 @@ npy_PyFile_DupClose2(PyObject *file, FILE* handle, npy_off_t orig_pos)
+        }
+        unbuf = PyObject_IsInstance(file, io_raw);
+        Py_DECREF(io_raw);
+        if (unbuf == 1) { // Unbuffered
```
See above for C++ comments and style.

charris commented:
LGTM modulo style. @pv, @mhvk Comments?

charris commented:
@cjermain Can you think of a way to test this?

cjermain commented:
I've fixed the C styling and squashed the changes.

I think the best way to test it is to construct a FileIO object and then modify the seek/tell methods to raise an error. Then run numpy.fromfile on it.

charris commented:
@cjermain Great. The proper place for a test looks to be `numpy/core/tests/test_multiarray.py`. Note that there are also some data files in `numpy/core/tests/data` if you want to add one for the original failure.

cjermain committed:
TST: for unbuffered IO without tell/seek

mhvk commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -193,9 +193,28 @@ npy_PyFile_Dup2(PyObject *file, char *mode, npy_off_t *orig_pos)
+        /* The io module is needed to determine if buffering is used */
+        io = PyImport_ImportModule("io");
+        if (io == NULL) {
+            return NULL;
```
Here and below, shouldn't the file handle be closed before returning `NULL`?

charris commented:
Looks like you merged master somewhere down the line, hence endolith's commit. I suspect that will go away if you update your master from upstream and rebase this branch on it.

cjermain commented:
Ok, I've rebased on the master which fixed the problem.

charris commented:
Merged. Thanks @cjermain .
